dns-policy-operator/src/charm.py

```diff
@@ -0,0 +1,208 @@
+
+    def dns_record_relations_data_to_entries(
+        self,
+        relation_data: list[
+            tuple[dns_record.DNSRecordRequirerData, dns_record.DNSRecordProviderData]
+        ],
+    ) -> list[models.DnsEntry]:
+        """Convert DNSRecordRequirerData to a list of DnsEntry.
+
+        Args:
+            relation_data: input relation data
+        Returns:
+            A list of DnsEntry
+        """
+        entries: list[models.DnsEntry] = []
+        for record_requirer_data, _ in relation_data:
+            for entry in record_requirer_data.dns_entries:
+                entries.append(models.create_dns_entry_from_requirer_entry(entry))
+        return entries
+
+    def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
+        """Handle changed configuration."""
+        self.unit.status = ops.MaintenanceStatus("Configuring workload")
+        self.dns_policy.configure(
+            {
+                "debug": "true" if self.config["debug"] else "false",
+                "allowed-hosts": json.dumps(
+                    [e.strip() for e in str(self.config["allowed-hosts"]).split(",")]
+                ),
+            }
+        )
+
+    def _on_start(self, _: ops.StartEvent) -> None:
+        """Handle start event."""
+
+    def _on_install(self, _: ops.InstallEvent) -> None:
+        """Handle install event."""
+        self.unit.status = ops.MaintenanceStatus("Preparing dns-policy-app")
+        self.dns_policy.setup(self.unit.name)
+        self._timer.start(self.unit.name, "reconcile", "30s", "1m")
+
+    def _on_database_created(self, _: DatabaseCreatedEvent) -> None:
+        """Handle database created.
+
+        Args:
+            event: Event triggering the database created handler.
+        """
+        database_relation_data = self._database.get_relation_data()
+        self.unit.status = ops.MaintenanceStatus("Preparing database")
+        self.dns_policy.configure(
+            {
+                "debug": "true" if self.config["debug"] else "false",
+                "allowed-hosts": json.dumps(
+                    [e.strip() for e in str(self.config["allowed-hosts"]).split(",")]
+                ),
+                "database-port": database_relation_data["POSTGRES_PORT"],
+                "database-host": database_relation_data["POSTGRES_HOST"],
+                "database-name": database_relation_data["POSTGRES_DB"],
+                "database-password": database_relation_data["POSTGRES_PASSWORD"],
+                "database-user": database_relation_data["POSTGRES_USER"],
+            }
+        )
+        self.dns_policy.command("migrate")
+
+    def _on_database_endpoints_changed(self, _: DatabaseEndpointsChangedEvent) -> None:
+        """Handle endpoints change.
+
+        Args:
+            event: Event triggering the endpoints changed handler.
+        """
+        database_relation_data = self._database.get_relation_data()
```

Comment:
minor/refactor: it looks the same to me so it could be extracted and both event handlers could call the same method.

Reply:
Refactored